### PR TITLE
Convert blob package to TypeScript

### PR DESCRIPTION
**Pull request by jpstevens**
## What?
This PR converts the `blob` package to TypeScript.

## Why?
Ensures package is fully type checked.

## How?
- converted files to `.ts`
- explicitly typed variables and functions in `index` file
- added `@ts-ignore` statements to tests that call functions in ways that are _not_ type-compliant

## Testing Instructions
`npm run test:unit -- packages/blob` tests pass
`npm run build:package-types` returns a zero exit code

### CODE REVIEW

File: packages/blob/src/index.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
+type BlobPart = BufferSource | Blob | string;
```
Comment (jsnajdr) [inline packages/blob/src/index.ts:1]: `BlobPart` should be a native built-in type, after all, it's used in the `Blob` constructor. You don't need to define it here.

File: packages/blob/src/index.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
+
 /**
  * @type {Record<string, File|undefined>}
  */
-const cache = {};
+const cache: Record< string, File | undefined > = {};
```
Comment (jsnajdr) [inline packages/blob/src/index.ts:6]: I think the cache type should be `Record< string, File >`. We never intentionally store `undefined` values.

Comment (jsnajdr) [reply packages/blob/src/index.ts:6]: Also, we can remove all of the JSDoc type annotations.

**Comment by sirreal** [comment]
👋 @jpstevens Thanks for this! If you'll update this PR on the latest `trunk` and [address feedback](https://github.com/WordPress/gutenberg/pull/61541#pullrequestreview-2054996395) we can likely give this a final review and land it.

**Pull request closed without merging**